logging/measure/tracer.go

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2018 Cisco and/or its affiliates.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package measure
+
+//go:generate protoc --proto_path=model/apitrace --gogo_out=model/apitrace model/apitrace/apitrace.proto
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ligato/cn-infra/logging"
+	"github.com/ligato/cn-infra/logging/measure/model/apitrace"
+)
+
+// Tracer allows to measure, store and list measured time entries.
+type Tracer interface {
+	// LogTime puts measured time to the table and resets the time.
+	LogTime(entity string, start time.Time)
+	// Get all trace entries stored
+	Get() *apitrace.Trace
+	// Clear removes entries from the log database
+	Clear()
+}
+
+// NewTracer creates new tracer object
+func NewTracer(name string, log logging.Logger) Tracer {
+	return &tracer{
+		name:   name,
+		log:    log,
+		index:  1,
+		timedb: make([]*entry, 0),
+	}
+}
+
+// Inner structure handling database and measure results
+type tracer struct {
+	sync.Mutex
+
+	name string
+	log  logging.Logger
+	// Entry index, used in database as key and increased after every entry. Never resets since the tracer object is
+	// created or the database is cleared
+	index uint64
+	// Time database, uses index as key and entry as value
+	timedb []*entry
+}
+
+// Single time entry
+type entry struct {
+	index      uint64
+	name       string
+	loggedTime time.Duration
+}
+
+func (t *tracer) LogTime(entity string, start time.Time) {
```

Comment:
This is good as very raw form of adding entries, but there should be an API creating entry which would store current time and then return it and later be able to call Log/Record for the entry to actually calculate duration and store the entry in `timedb`. 

This allows users to not have to take care of their start time. And more importantly this will store the index at the point of entry creation and not when it gets logged. That way all indexes for entries will be ordered chronologically just like they started.